src/connection-secure.c

```diff
@@ -0,0 +1,547 @@
+/*
+ * Punica - LwM2M server with REST API
+ * Copyright (C) 2019 8devices
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "connection-secure.h"
+
+#define BUFFER_SIZE 1024
+
+typedef struct secure_connection_context_t
+{
+    connection_api_t api;
+    device_connection_t *connection_list;
+    int port;
+    int address_family;
+    const char *certificate_file;
+    const char *private_key_file;
+    int listen_socket;
+    gnutls_certificate_credentials_t server_cert;
+    gnutls_priority_t priority_cache;
+    gnutls_datum_t cookie_key;
+    gnutls_psk_server_credentials_t server_psk;
+    void *data;
+    f_psk_cb_t psk_cb;
+} secure_connection_context_t;
+
+static ssize_t prv_net_send(gnutls_transport_ptr_t context, const void *data, size_t size)
+{
+    device_connection_t *conn = (device_connection_t *)context;
+
+    return sendto(conn->sock, data, size, 0, (struct sockaddr *)&conn->addr, conn->addr_size);
+}
+
+static int prv_new_socket(void *this)
```

Comment:
IMO even though this is C and not C++, reserved keywords should still be avoided when possible. Also I don't see a reason for this and several other static functions to use `void *` type. In general void types should only be used when actively hiding information from other layers, such as in public-facing API functions.